File: src/CopilotDiscordBot/Utils/TextChunker.cs
namespace CopilotDiscordBot.Utils;

public static class TextChunker
{
    public static IEnumerable<string> ChunkForDiscord(string text, int maxLen = 1900)
    {
        if (string.IsNullOrEmpty(text)) yield break;
        text = text.Replace("\r\n", "\n");

        var start = 0;
        while (start < text.Length)
        {
            var len = Math.Min(maxLen, text.Length - start);

            // Prefer splitting on newline near the end.
            var split = text.LastIndexOf('\n', start + len - 1, len);
            if (split <= start + 100) split = start + len; // avoid tiny chunks

            var part = text.Substring(start, split - start).TrimEnd();
            if (part.Length > 0) yield return part;

            start = split;
            while (start < text.Length && text[start] == '\n') start++;
        }
    }
}

File: src/utils/textChunker.ts
export function* chunkForDiscord(text: string, maxLen = 1900): Generator<string> {
  if (!text) return;
  text = text.replace(/\r\n/g, "\n");

  let start = 0;
  while (start < text.length) {
    const len = Math.min(maxLen, text.length - start);
    let split = text.lastIndexOf("\n", start + len - 1);

    // Avoid tiny leading chunks
    if (split <= start + 100) {
      split = start + len;
    }

    const part = text.slice(start, split).trimEnd();
    if (part.length > 0) yield part;

    start = split;
    while (start < text.length && text[start] === "\n") start++;
  }
}
